### ISSUE TITLE
Major efficiency improvements from building from source (not mentioned in docs)

### ISSUE THREAD

**Issue by DiffeoInvariant** [google/jax#5730]
Hi JAX devs,

One thing I've observed recently is massive performance gains (2x speed increase or more) from building `jaxlib` from source, likely as a result of XLA's ops being compiled with all the vector instructions my machine has (as opposed to the binary that `pip` downloads). I first noticed this after building the (ROCm) GPU version and noticing that it was a lot faster than the pip-installed CPU only version even though my CPUs have higher floating-point bandwidth than my GPU, so I tried comparing the CPU-only pip version and the build-from-source CPU-only version, and found the massive performance gains I'm talking about here. 

This doesn't seem to be mentioned anywhere in the JAX documentation, and it would probably be useful for a lot of people, so I figured I'd bring it up here. I'm happy to add a note in the docs myself and make a PR if that's easier for y'all than writing it yourselves and you'd want that.

I suppose you could also compile multiple variants of `jaxlib` and include those with the pip package if you think that's a better solution (to just make the pip package fast on more systems)...

**Comment by hawkinsp** [comment]
This is actually a little bit surprising to me!

The reason I say this is that JAX (via XLA) is a just-in-time compiler. That is, it generates code for your particular hardware, tailored for your particular hardware, every time it is run. For example, if you have an Intel Skylake CPU, JAX would generate AVX512 instructions via JIT compilation, even though the `jaxlib` wheel itself does not require or use such instructions. The build flags should only really affect things like the speed of the compiler itself, and should not benefit significantly from, say, vectorization.

Can you share a benchmark and instructions to reproduce the issue? I'd like to reproduce this so I can dig into it a little bit deeper and see what is going on.

(As it happens, I was thinking of at the very least requiring AVX instructions for JAX shortly for a host kernel I'm planning to add, but I'm surprised there's any meaningful difference at the moment!)

**Comment by DiffeoInvariant** [comment]
Sure, I'll put together a good benchmark for this (the code I observed the effect on isn't a good benchmark, but I should be able to extract a good one from it) and drop it here later today. 

Meanwhile, are you sure that XLA actually compiles _everything_ at runtime? For example, when I was building `jaxlib`, I saw that files like `org_tensorflow/tensorflow/compiler/xla/service/cpu/runtime_matmul.cc` were being compiled. This caused me to think that maybe XLA pre-compiles certain ops (e.g. matrix multiplication) and links them in at runtime for faster compiles---is this true?  If so, it would explain the difference, I think. If not, why are those files being compiled at all?

**Comment by hawkinsp** [comment]
There are a small number of kernels (matmul, convolution) that are precompiled. It's possible they are relevant here.

That said, most 32-bit kernels use MKL-DNN, which I believe also uses JIT compilation internally (https://oneapi-src.github.io/oneDNN/dev_guide_cpu_dispatcher_control.html). There are however a number of Eigen kernels that are not dispatched at runtime; it's possible you are hitting one of them. So a benchmark would be helpful.

**Comment by DiffeoInvariant** [comment]
I'm re-building JAX in a fresh environment now to test out the benchmark I've written, so as soon as that's done and I've verified that the benchmark shows what I think it will (probably in about an hour based on the current compilation rate) I'll drop it and instructions for replication here. Based on what's going on in the actual code I observed this effect in, I'll be surprised if it doesn't turn out that matrix multiplication kernels explain almost the entire difference, since the majority of the FLOPS are spent on a few big batched matrix multiplies (i.e. multiplying a whole bunch of 3x3 matrices against each other by `vmap`-ing over axis 0 of a 3-d array)

**Comment by hawkinsp** [comment]
What are the types of the matrices? Are they float64 or float32?

**Comment by DiffeoInvariant** [comment]
float64

**Comment by DiffeoInvariant** [comment]
FWIW, I'll be running the benchmark on a Haswell CPU, so I have AVX2 but not AVX512

**Comment by DiffeoInvariant** [comment]
Hm, so this is somewhat interesting; the difference I observed must be in something other than matrix multiplication, because the following benchmark shows about a 10% time reduction from using locally-compiled jaxlib as opposed to pip-installed jaxlib on my machine (as opposed to the ~50% improvement I saw on my production code):

```
import jax
from jax import numpy as jnp
from jax import random
from jax.config import config
from sys import argv
import time
config.update('jax_enable_x64', True)


N = 5000
if len(argv) > 1:
    N = int(argv[1])


def time_execution(f, args):
    start = time.time()
    out = f(*args).block_until_ready()
    stop = time.time()
    print(f"Execution time (seconds): {stop - start}")
    return stop - start

@jax.jit
@jax.vmap
def func(x):
    xx = x @ x
    xxx = xx @ x
    return jnp.array([jnp.trace(x), jnp.trace(xx), jnp.trace(xxx)])

key = random.PRNGKey(0)
key, subkey = random.split(key)
A = random.normal(key, shape=(N,3,3))

ttot = 0.0
for _ in range(10):
    ttot += time_execution(func, (A,))

print(f"average time: {ttot / 10}")
```


I'll try to come up with a similarly compact benchmark that reproduces the more dramatic speed improvement I originally observed on my code.

**Comment by hawkinsp** [comment]
Float64 matmul and/or convolution could explain things. MKL-DNN (aka one-DNN) only supports float32. For other cases, we use Eigen, which does depend on the target architecture at build time. Unfortunately it is somewhat difficult to safely mix Eigen kernels built with different compilation options in the same binary without creating ODR problems.

The question I have is: would requiring AVX support (as TensorFlow's binaries do already) be enough to capture most of the performance benefit? To answer that, we'd need a good benchmark, and that's where you can help!

**Comment by DiffeoInvariant** [comment]
I see --- I'll let you know when I have a better benchmark than the one above (i.e. one that shows more than 20%-ish of the speed difference that I observed, which is all I captured in that one). Thanks!

**Comment by hawkinsp** [comment]
#5738 enables AVX by default. I'd be interested to know if it help significantly, although I can probably verify that myself if you have a benchmark.

**Comment by DiffeoInvariant** [comment]
Great -- I'll test that either later today or tomorrow morning. I tried unsuccessfully to reproduce the problem in a reasonably short (lines of code-wise) benchmark earlier (I could only reproduce an approximately 10% speed difference as opposed to the 50% I saw originally, and that was pretty consistent across the several variations of benchmark program I tried), but I'm interested in figuring out what's going on here so I'll try a bit more tomorrow to reproduce it in a benchmark I can share

**Comment by hawkinsp** [comment]
Closing; I'm assuming that enabling AVX by default is enough to fix this. Please reopen if you have a benchmark that proves otherwise!
